README.md
# gh-projects-discord

Relays `GitHub Projects` webhooks to `Discord`. Written in Haskell to have fun.

## Existing GitHub-Discord integration

Using the `/discord` endpoint, most common webhooks like "issue created" or "commit pushed" get shown as Discord embeds in the directed channel. However, webhooks related to GitHub Projects are entirely ignored. This may be because the feature is still new, however it considerably limits the adoption of Projects as a central kanban board in team projects.

I decided it would help a lot if I build a relayer myself, basically acting as a proxy for webhooks.

## Flow

1. GitHub pushes a webhook event to the server running `gh-projects-discord-exe` (port 3000).
2. The service parses and optionally fetches more information relating to the Project board.
3. The service pushes a webhook event to Discord using the webhook URL passed in as the cli argument.

![Demo Image](https://i.imgur.com/OfVpXv1.png)

## Run

1. Create a `gh.token` file containing `<github-username>:<access token>`
2. Run `stack run -- gh-projects-discord-exe "https://discord/webhook/url..."`

## License

[MIT](./LICENSE)


ChangeLog.md
# Changelog for gh-projects-discord

## 0.2.0

- Now returns 202 on ignored webhooks
- Now requests GH for issue title/link if the Project Card is an issue, instead of showing `(empty)`

## 0.1.0

- Completed a functional relayer for most Project events


src/DiscordSender.hs
{-# LANGUAGE RecordWildCards #-}

module DiscordSender
    ( sendEmbed
    , Embed(..)
    , EmbedAuthor(..)
    ) where


import           Data.Aeson
import qualified Data.ByteString.Lazy as LBS
import qualified Data.Text as T
import           Network.HTTP.Conduit
import           Network.HTTP.Types.Status

-- | ADT for the Discord Webhook Execute request body.
data Params = Params { content  :: T.Text
                     , username :: T.Text
                     , avatar_url :: T.Text
                     , tts      :: Bool
                     , embeds   :: [Embed]
                     }

instance ToJSON Params where
    toJSON Params{..} = object
        [ "content" .= content
        , "username" .= username
        , "avatar_url" .= avatar_url
        , "tts" .= tts
        , "embeds" .= embeds
        ]

-- | ADT for the Discord Embed type. Unused fields are not implemented.
data Embed = Embed { embedTitle :: T.Text
                   , embedType :: T.Text
                   , embedDescription :: T.Text
                   , embedUrl :: T.Text
                   , embedColor :: Integer
                   , embedAuthor :: EmbedAuthor
                   } 

instance ToJSON Embed where
    toJSON Embed{..} = object
        [ "title" .= embedTitle
        , "type" .= embedType
        , "description" .= embedDescription
        , "url" .= embedUrl
        , "color" .= embedColor
        , "author" .= embedAuthor
        ]

-- | ADT for the Discord Embed Author type.
data EmbedAuthor = EmbedAuthor { embedAuthorName :: T.Text
                               , embedAuthorUrl :: T.Text
                               , embedAuthorIcon :: T.Text
                               }

instance ToJSON EmbedAuthor where
    toJSON EmbedAuthor{..} = object
        [ "name" .= embedAuthorName
        , "url" .= embedAuthorUrl
        , "icon_url" .= embedAuthorIcon
        ]

-- | Send a POST request to a webhook endpoint, with the Discord params.
sendRequest :: String -> Params -> IO ()
sendRequest endpoint params = do
    initRequest <- parseRequest endpoint
    let request = initRequest { method = "POST"
                              , requestBody = RequestBodyLBS $ encode params
                              , requestHeaders = [("Content-Type", "application/json")]
                              }
    manager <- newManager tlsManagerSettings
    res <- httpLbs request manager
    case () of
        _ | responseStatus res == status204 -> pure ()
        _                                   -> LBS.putStr $ responseBody res

-- | Wrapper for sendRequest, specifically for sending embeds.
sendEmbed :: String -> Embed -> IO ()
sendEmbed endpoint embed = do
    let params = Params { content = ""
                        , username = "GitHub"
                        , avatar_url = "https://github.githubassets.com/images/modules/logos_page/GitHub-Mark.png"
                        , tts = False
                        , embeds = [embed]
                        }
    sendRequest endpoint params


src/GitHubWebhookDecoder.hs
module GitHubWebhookDecoder
    ( Webhook(..)
    , Item(..)
    , Context(..)
    , Action(..)
    , parseWebhook
    , parseBoardName
    , parseBoardUrl
    , parseBoardBody
    , parseColumnName
    , parseCardNote
    , parseCardColumnUrl
    , parseCardIssueUrl
    , parseProjectUrl
    , parseIssueTitle
    , parseIssueUrl
    ) where

import           Data.Aeson
import           Data.Aeson.Types
import           Data.Foldable          ( asum )
import qualified Data.HashMap.Strict as HM
import qualified Data.Text as T

-- | ADT for the Webhook as a whole.
data Webhook = Webhook Context Action Item deriving Show

-- | ADT for the value inside the webhook.
data Item = Project Value
          | ProjectCard Value
          | ProjectColumn Value
          deriving Show

-- | ADT for the context in which the action was performed in.
data Context = Context  { senderName    :: T.Text
                        , senderImage   :: T.Text
                        , repoName      :: T.Text
                        , repoOwnerName :: T.Text
                        }
                        deriving Show

-- | ADT for the action verb that was performed on an item.
data Action = Created
            | Edited
            | Moved
            | Closed
            | Reopened
            | Deleted
            deriving Show


-- | Parses a decoded JSON Value to a Webhook.
parseWebhook :: Value -> Parser Webhook
parseWebhook v = do
    context <- parseContext v
    action <- parseAction v
    item <- parseItem v
    pure $ Webhook context action item

-- | Parses a decoded JSON Value to a Context.
parseContext :: Value -> Parser Context
parseContext = withObject "entire webhook" $ \o -> do
    sender <- o .: "sender"
    senderName <- sender .: "login"
    senderImage <- sender .: "avatar_url"
    repository <- o .: "repository"
    repoName <- repository .: "name"
    repoOwner <- repository .: "owner"
    repoOwnerName <- repoOwner .: "login"
    pure $ Context senderName senderImage repoName repoOwnerName

-- | Parses a decoded JSON Value to an Action.
parseAction :: Value -> Parser Action
parseAction = withObject "entire webhook" $ \o -> do
    action <- o .: "action" :: Parser String
    case action of
        "created"   -> pure Created
        "edited"    -> pure Edited
        "moved"     -> pure Moved
        "closed"    -> pure Closed
        "reopened"  -> pure Reopened
        "deleted"   -> pure Deleted
        _           -> fail "Action not supported."

-- | Parses a decoded JSON Value to an Item.
parseItem :: Value -> Parser Item
parseItem = withObject "entire webhook" $ \o ->
    asum [ ProjectCard <$> (o .: "project_card")
         , ProjectColumn <$> (o .: "project_column")
         , Project <$> (o .: "project")
         ]

-- | Parses a Project to its name.
parseBoardName :: Value -> Parser T.Text
parseBoardName = withObject "project board" $ \o -> o .: "name"

-- | Parses a Project to its URL.
parseBoardUrl :: Value -> Parser T.Text
parseBoardUrl = withObject "project board" $ \o -> o .: "html_url"

-- | Parses a Project to its body text.
parseBoardBody :: Value -> Parser T.Text
parseBoardBody = withObject "project board" $ \o -> o .: "body"

-- | Parses a Column to its name.
parseColumnName :: Value -> Parser T.Text
parseColumnName = withObject "project board column" $ \o -> o .: "name"

-- | Parses a Card to its contents.
parseCardNote :: Value -> Parser T.Text
parseCardNote = withObject "project board card" $ \o -> o .: "note"

-- | Parses a Card to its column API URL.
parseCardColumnUrl :: Value -> Parser T.Text
parseCardColumnUrl = withObject "project board card" $ \o -> o .: "column_url"

-- | Parses a Card to its issue API Url.
parseCardIssueUrl :: Value -> Parser T.Text
parseCardIssueUrl = withObject "project board card" $ \o -> o .: "content_url"

-- | Parses a Column or a Card to its project API URL.
parseProjectUrl :: Value -> Parser T.Text
parseProjectUrl = withObject "project board column or card" $ \o -> o .: "project_url"

-- | Parses an Issue to its title.
parseIssueTitle :: Value -> Parser T.Text
parseIssueTitle = withObject "issue" $ \o -> o .: "title"

-- | Parses an Issue to its URL.
parseIssueUrl :: Value -> Parser T.Text
parseIssueUrl = withObject "issue" $ \o -> o .: "html_url"


src/GitHubFurtherRequester.hs
module GitHubFurtherRequester
    ( Token
    , requestFurther
    ) where

import           Data.Aeson
import qualified Data.ByteString.Lazy as LBS
import qualified Data.Text as T
import qualified Data.Text.Encoding as TE
import           Network.HTTP.Conduit
import           Network.HTTP.Types.Status

type Token = T.Text

-- | Perform a GET request to the specified GitHub API url, using the
-- token of format "user:token". Returns a decoded Aeson Value.
requestFurther :: Token -> T.Text -> IO Value
requestFurther token url = do
    let [user, password] = T.splitOn ":" token
    initReq <- parseRequest $ T.unpack url
    let initReq' = initReq { requestHeaders = [("Accept", "application/vnd.github.inertia-preview+json")
                                              ,("User-Agent", "yutotakano/gh-projects-discord")
                                              ]
                           }
    let req = applyBasicAuth (TE.encodeUtf8 user) (TE.encodeUtf8 password) initReq'
    manager <- newManager tlsManagerSettings
    res <- httpLbs req manager

    case eitherDecode (responseBody res) of
        Left x  -> fail "GitHub API Response Error"
        Right x -> pure x



src/Lib.hs
module Lib
    ( main
    ) where

import           Control.Monad.IO.Class     ( liftIO )
import           Data.Aeson
import           Data.Aeson.Types           ( parseEither
                                            , parseMaybe
                                            , Parser(..)
                                            )
import           Data.Char                  ( isSpace )
import qualified Data.Text.Lazy as TL
import qualified Data.Text as T
import qualified Data.Text.IO as TIO
import qualified Network.HTTP.Types.Status as Status
import           System.Environment
import           Web.Scotty

import           DiscordSender
import           GitHubWebhookDecoder
import           GitHubFurtherRequester

-- | Entry point.
main :: IO ()
main = do
    args <- getArgs
    webhookUrl <- case () of
        _ | length args == 0 -> error "URL not supplied"
        _                    -> pure $ head args
    tokenContents <- TIO.readFile "gh.token" :: (IO Token)
    let token = T.takeWhile (not . isSpace) tokenContents
    
    scotty 3000 $
        post "/github" $ do
            webhook <- decodeWebhook
            liftIO $ createEmbed token webhook >>= sendEmbed webhookUrl
            html $ "Success."

-- | Decode the JSON in the request body to a Webhook.
-- Silently fails if the body is empty or unparsable.
decodeWebhook :: ActionM Webhook
decodeWebhook = do
    body <- body
    decoded <- case eitherDecode body of
        Left x  -> silentFail (TL.pack x)
        Right x -> pure x

    let webhook = parseEither parseWebhook decoded
    case webhook of
        Left x  -> silentFail (TL.pack x)
        Right x -> pure x

-- | End request with a 202. Used to show success, but nothing acted.
silentFail :: TL.Text -> ActionM a
silentFail t = do
    status Status.status202
    text t
    finish

-- | Create an Embed in the IO Monad, using the GH token
-- and the Webhook data.
createEmbed :: Token -> Webhook -> IO Embed
createEmbed t (Webhook c a (Project v)) = do
    let boardName = parseMaybeDef parseBoardName "Untitled Board" v
    let boardUrl = parseMaybeDef parseBoardUrl "" v
    let boardBody = parseMaybeDef parseBoardBody "(empty)" v
    let title = titlePrefix c <> "Project Board " <> (T.pack $ show a) <> ": " <> boardName
    let description = "Note: `" <> boardBody <> "`"
    pure $ (def c a)
        { embedTitle = title
        , embedDescription = description
        , embedUrl = boardUrl
        }

createEmbed t (Webhook c a (ProjectColumn v)) = do
    let columnName = parseMaybeDef parseColumnName "" v
    (boardName, boardUrl) <- getBoard t v
    let title = titlePrefix c <> "Project Column " <> (T.pack $ show a) <> ": " <> boardName
    let description = "Column " <> (T.pack $ show a) <> ": `" <> columnName <> "`"
    pure $ (def c a)
        { embedTitle = title
        , embedDescription = description
        , embedUrl = boardUrl
        }

createEmbed t (Webhook c Moved (ProjectCard v)) = do
    -- Handle move event differently
    cardNote <- getCardNote t v

    let columnApiUrl = parseMaybeDef parseCardColumnUrl "" v
    column <- requestFurther t columnApiUrl
    let columnName = parseMaybeDef parseColumnName "" column
    (boardName, boardUrl) <- getBoard t v
    let title = titlePrefix c <> "Project Card Moved: " <> boardName
    let description = "Moved card " <> cardNote <> " to `" <> columnName <> "`"
    pure $ (def c Moved)
        { embedTitle = title
        , embedDescription = description
        , embedUrl = boardUrl
        }

createEmbed t (Webhook c a (ProjectCard v)) = do
    cardNote <- getCardNote t v
    (boardName, boardUrl) <- getBoard t v
    let title = titlePrefix c <> "Project Card " <> (T.pack $ show a) <> ": " <> boardName
    let description = "Card " <> (T.pack $ show a) <> ": " <> cardNote
    pure $ (def c a)
        { embedTitle = title
        , embedDescription = description
        , embedUrl = boardUrl
        }

-- | Create a "[repo/user]" prefix for all embed titles.
titlePrefix :: Context -> T.Text
titlePrefix c = "[" <> repoOwnerName c <> "/" <> repoName c <> "] "

-- | Run a parser with a default value if it fails.
parseMaybeDef :: (a -> Parser b) -> b -> a -> b
parseMaybeDef parser def v  =
    case parseMaybe parser v of
        Nothing -> def
        Just x  -> x

-- | Request GH for a board object through the preview GitHub Project API.
getBoard :: Token -> Value -> IO (T.Text, T.Text)
getBoard token v = do
    let projectApiUrl = parseMaybeDef parseProjectUrl "" v
    projectBoard <- requestFurther token projectApiUrl
    let projectBoardName = parseMaybeDef parseBoardName "Untitled Board" projectBoard
    let projectBoardUrl = parseMaybeDef parseBoardUrl "" projectBoard

    pure (projectBoardName, projectBoardUrl)

-- | Parse the card note, or if it's an Issue, request GH for an issue object
-- and get the title and link.
getCardNote :: Token -> Value -> IO T.Text
getCardNote t v = case parseMaybe parseCardNote v of
    Just x  -> pure $ "`" <> x <> "`"
    Nothing -> do
        -- Attempt to grab the relevant issue
        let mbIssueApiUrl = parseMaybe parseCardIssueUrl v
        case mbIssueApiUrl of
            Nothing -> pure "(empty)"
            Just issueApiUrl  -> do
                issue <- requestFurther t issueApiUrl
                let issueTitle = parseMaybeDef parseIssueTitle "" issue
                let issueUrl = parseMaybeDef parseIssueUrl "" issue
                pure $ "[`" <> issueTitle <> "`](" <> issueUrl <> ")"

-- | Default ADT for Embed.
def :: Context -> Action -> Embed
def c a = Embed { embedTitle = ""
                , embedUrl = ""
                , embedDescription = ""
                , embedType = "rich"
                , embedColor = colorFromAction a
                , embedAuthor = createAuthor c
                }

-- | Transform Actions to decimal color codes.
colorFromAction :: Action -> Integer
colorFromAction Created = 6667344 -- #65bc50
colorFromAction Edited = 16766784 -- #ffd740
colorFromAction Moved = 16766784 
colorFromAction Closed = 0
colorFromAction Reopened = 11771355 -- #b39ddb
colorFromAction Deleted = 6588634 -- #6488da

-- | Transform Context to an EmbedAuthor for Embeds.
createAuthor :: Context -> EmbedAuthor
createAuthor context = EmbedAuthor { embedAuthorName = senderName context
                                   , embedAuthorUrl = "https://github.com/" <> senderName context
                                   , embedAuthorIcon = senderImage context
                                   }
